rust/windows-client/src-tauri/src/client/crash_handling.rs

```diff
@@ -0,0 +1,172 @@
+
+    #[allow(unsafe_code)]
+    let handler = crash_handler::CrashHandler::attach(unsafe {
+        crash_handler::make_crash_event(move |crash_context: &crash_handler::CrashContext| {
+            // Before we request the crash, send a message to the server
+            client.send_message(2, "mistakes were made").unwrap();
+
+            // Send a ping to the server, this ensures that all messages that have been sent
+            // are "flushed" before the crash event is sent. This is only really useful
+            // on macos where messages and crash events are sent via different, unsynchronized,
+            // methods which can result in the crash event closing the server before
+            // the non-crash messages are received/processed
+            client.ping().unwrap();
+
+            crash_handler::CrashEventResult::Handled(client.request_dump(crash_context).is_ok())
+        })
+    })
+    .context("failed to attach signal handler")?;
+
+    // On linux we can explicitly allow only the server process to inspect the
+    // process we are monitoring (this one) for crashes
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    {
+        handler.set_ptracer(Some(_server.id()));
+    }
+
+    Ok(handler)
+}
+
+fn connect() -> anyhow::Result<(Client, std::process::Child)> {
+    let exe = std::env::current_exe().context("unable to find our own exe path")?;
+    let mut server = None;
+
+    for _ in 0..10 {
+        // Create the crash client first so we can error out if another instance of
+        // the Firezone client is already using this socket for crash handling.
+        if let Ok(client) = Client::with_name(SOCKET_NAME) {
+            return Ok((
+                client,
+                server.ok_or_else(|| {
+                    anyhow!(
+                        "should be impossible to make a client if we didn't make the server yet"
+                    )
+                })?,
+            ));
+        }
+
+        server = Some(
+            std::process::Command::new(&exe)
+                .arg("crash-handler-server")
+                .spawn()
+                .context("unable to spawn server process")?,
+        );
+
+        // Give it time to start
+        std::thread::sleep(std::time::Duration::from_millis(100));
+    }
+
+    bail!("Couldn't set up crash handler server")
+}
+
+/// Main function for the server process, for out-of-process crash handling.
+///
+/// The server process seems to be the preferred method,
+/// since it's hard to run complex code in a process
+/// that's already crashed and likely suffered memory corruption.
+///
+/// <https://jake-shadle.github.io/crash-reporting/#implementation>
+/// <https://chromium.googlesource.com/breakpad/breakpad/+/master/docs/getting_started_with_breakpad.md#terminology>
+pub(crate) fn server() -> anyhow::Result<()> {
+    let mut server = Server::with_name(SOCKET_NAME)?;
+
+    let ab = std::sync::atomic::AtomicBool::new(false);
+
+    struct Handler;
```

Comment:
rather than defining it inside a function let's do a submodule?

Reply:
I extracted it from the function, but the whole crash handling module is only 165 lines so I didn't split it yet.